emails/tests/views_tests.py

```diff
@@ -1270,36 +1315,406 @@ def test_complaint_disables_mask(self):
+
+        with self.assertLogs(INFO_LOG) as logs:
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        premium_user.profile.refresh_from_db()
+        assert premium_user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (rec1, rec2) = logs.records
+        assert rec1.msg == "_handle_complaint: developer_mode"
+        assert getattr(rec1, "mask_id") == domain_address.metrics_id
+        assert getattr(rec1, "dev_action") == "log"
+
+        assert rec2.msg == "complaint_notification"
+
+    def test_complaint_no_complained_recipients_error_logged(self):
+        """Without complained recipients, an error is logged, but matches on From"""
+        complaint_msg = deepcopy(self.complaint_msg)
+        del complaint_msg["complaint"]["complainedRecipients"]
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with (
+            self.assertLogs(INFO_LOG) as info_logs,
+            self.assertLogs(ERROR_LOG) as error_logs,
+        ):
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (info_log,) = info_logs.records
+        assert info_log.msg == "complaint_notification"
+        assert getattr(info_log, "user_match") == "found"
+        assert getattr(info_log, "relay_action") == "auto_block_spam"
+        assert getattr(info_log, "found_in") == "from_header"
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Unexpected message"
+        assert getattr(err_log, "missing_key") == "complainedRecipients"
+        assert getattr(err_log, "found_keys") == (
+            "arrivalDate,complaintFeedbackType,feedbackId,timestamp,userAgent"
+        )
+
+    def test_complaint_empty_complained_recipients_error_logged(self):
+        """With empty complained recipients, an error is logged, but matches on From"""
+        complaint_msg = deepcopy(self.complaint_msg)
+        complaint_msg["complaint"]["complainedRecipients"] = []
+        complaint_body = {"Message": json.dumps(complaint_msg)}
+
+        with self.assertLogs(ERROR_LOG) as error_logs:
+            response = _sns_notification(complaint_body)
+        assert response.status_code == 200
+
+        self.user.profile.refresh_from_db()
+        assert self.user.profile.auto_block_spam is True
+        self.mock_ses_client.send_raw_email.assert_not_called()
+
+        (err_log,) = error_logs.records
+        assert err_log.msg == "_get_complaint_data: Empty complainedRecipients"
+
+    def test_complaint_wrong_complained_recipients_error_logged(self):
+        """With an unexpected object in complained recipients, an error is logged"""
```

Comment:
same suggestion (non-blocking):

```suggestion
        """
        With unexpected value in complainedRecipients,
        we can still match the user on From:, perform the appropriate action,
        and log an error
        """
```